[Ng Wee Han] iP

beefham opened the pull request:
# Duke + A bit more
Duke has been upgraded to manage your schedule. These include:
* Your things **to do** anytime
* Your ~fearsome~ **deadlines**
* and your **events** that are happening

Then you simply just:
1. Add tasks using commands like `todo` or `event`
2. Tasks show up in Duke (use `list` to view them)
3. ❓ ❔ ❓ 
4. Profit

## Made up review
> Best app [10/10](https://www.youtube.com/watch?v=dQw4w9WgXcQ).
> -- Anonymous

### Features
- [x] Manage tasks
- [x] Manage deadlines
- [x] Manage events
- [x] Save to disk
- [ ] ~Play games~

### Sample code
``` Java
// please use Duke
if (!user.isUsingDuke) {
    user.setIsUsingDuke(true);
}
```

Commit:
Add Gradle support

beefham committed:
Add skeletal version of Duke

beefham committed:
Add task list to Duke

beefham committed:
add ability to mark tasks as complete

beefham committed:
add todo, deadline and event types

beefham committed:
Add exception catching and help dialog

beefham committed:
add ability to delete tasks

beefham committed:
add semi-automated testing

beefham committed:
add ability to read from file

beefham committed:
add ability to write to file

beefham committed:
add date parsing to deadline

beefham committed:
Merge branch 'branch-Level-7'

beefham committed:
resolve deadline due date type conflict

beefham committed:
refactor using more OOP, part 1: Level 1 to 3

beefham committed:
refactor using more OOP, part 2: Level 4 to 8

beefham committed:
refactor files by putting them in packages

beefham committed:
add simple junit tests

beefham committed:
create jar file

beefham committed:
add javadocs to duke and tasks

beefham committed:
edit to follow coding standard

beefham committed:
add ability to find tasks

beefham committed:
Merge branch 'branch-A-CodingStandard'

beefham committed:
fix merge conflict

beefham committed:
fix merge conflict

AllardQuek commented on src/main/java/commands/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,29 @@
+
+import duke.*;
+import tasks.*;
+import exceptions.*;
```
Should the specific imports be used here instead?

AllardQuek commented on src/main/java/commands/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Override
+    public void execute(Ui ui, TaskList taskList, Storage storage) {
+        Deadline newDeadline = new Deadline(this.desc, this.isDone, this.deadline);
```
Might be nit-picking, but you could consider not using the `this` keyword here? 

My tutor mentioned mentioned to avoid unnecessary usage of `this`, but I would still also prefer using `this` haha

AllardQuek commented on src/main/java/commands/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,29 @@
+
+import java.io.IOException;
+import java.time.LocalDate;
+
```
I like how the different types of imports are separated here 👍

AllardQuek commented on src/main/java/commands/AddEventCommand.java:
```diff
@@ -0,0 +1,29 @@
+import exceptions.*;
+
+public class AddEventCommand extends AddTaskCommand {
+
```
Not sure if this is an unnecessary blank line here?

AllardQuek commented on src/main/java/commands/AddEventCommand.java:
```diff
@@ -0,0 +1,29 @@
+
+import java.io.IOException;
+
+import duke.*;
+import tasks.*;
+import exceptions.*;
```
I notice the same issue above and in other files as well

AllardQuek commented on src/main/java/commands/AddTaskCommand.java:
```diff
@@ -0,0 +1,36 @@
+        }
+    }
+
+    public boolean isQuit() { return false; }
```
Should the body be on multiple lines instead? I think it's a coding standard for if-else statements, but not sure about methods. Could be easier to debug when line numbers come into play 🤔

AllardQuek commented on src/main/java/commands/MarkDoneCommand.java:
```diff
@@ -0,0 +1,40 @@
+        }
+    }
+
+    public boolean isQuit() {return false;}
```
Should there be spacing between the braces and the method body?

AllardQuek commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+
+    private Ui ui;
+    private TaskList taskList;
+    private Storage store;
```
Would `store` be a bit vague since `store` might have multiple meanings? (textbook page 62) 

Feel free to ignore !

AllardQuek commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+    private Storage store;
+    private boolean isRunning;
+
+    Duke() {
```
LGTM!

AllardQuek commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,130 @@
+package duke;
+
+import java.io.File;
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
```
Perhaps the imports could be ordered alphabetically?

beefham committed:
Merge branch 'A-Gradle'

AllardQuek commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,130 @@
+        try (BufferedReader br = new BufferedReader(new FileReader(this.taskFile))) {
+            ArrayList<Task> taskList = new ArrayList<>();
+            String line;
+            while ((line = br.readLine()) != null && !line.equals("")) {
```
Not sure if there is a clearer way to express this? Not sure if it's considered a complicated expression (textbook: page 57-58)

AllardQuek commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    /**
+     * Prints out the message when a task has been successfully added to the TaskList in Duke.
+     * @param t The task that was added to the TaskList.
```
Maybe add a new line after the header?

AllardQuek reviewed (commented):
I realised we should only comment on coding standards but some points here are more of code quality and are quite nit-picky. Looks good overall!

beefham committed:
add gradle support

beefham committed:
add checkstyle

beefham committed:
fix code that did not follow the coding standard

beefham committed:
add GUI to Duke up to tutorial 3

beefham committed:
create working gui for duke

beefham committed:
add gui to duke

beefham committed:
Merge branch 'branch-Level-10'

danielsimre commented on src/main/java/tasks/Deadline.java:
```diff
@@ -0,0 +1,32 @@
+
+public class Deadline extends Task {
+
+    LocalDate dueDate;
```
I noticed that in a lot of classes, many class variables are missing their access modifier. Perhaps you should explicitly state these as private? (Unless they are meant to be package private)

danielsimre commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,61 @@
+        this.isDone = false;
+    }
+
+    String completionStatus() {
```
Should the access modifier for this function be explicitly stated? (Either public or protected, from what I see of its usage)

danielsimre commented on src/main/java/duke/DukeParser.java:
```diff
@@ -0,0 +1,76 @@
+        } catch (Exception e) { //TODO DANGEROUS! please change
+            return new ErrorCommand("An exception occurred!");
+        }
+    }
```
Currently this function is about 50 lines long, so perhaps you could find a way to reduce it to about 30 lines? (One possible way would be for the caller of `parse` to contain the try-catch block to handle errors)

danielsimre commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,63 @@
+ *
+ */
+public class TaskList {
+    private ArrayList<Task> taskList;
```
Maybe this variable could be called `tasks` instead since this variable is a collection of tasks?

danielsimre reviewed (commented):
Overall, the quality of your code is good! I only managed to find a few minor errors, plus a few that Allard has already caught. (Explicit import statements and usage of `this` keyword)

stephanie-csy commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,104 @@
+
+import java.io.File;
+
+import commands.*;
+import tasks.*;
```
Should import everything explicitly instead of using wildcard import

stephanie-csy reviewed (commented):
Everything looks neat and readable, cannot find any major coding standards violations. Maybe you can add javadocs for all public classes and methods so that others reading your code can easily understand what everything does at one glance. Also, consider naming your packages `duke.task`, `duke.command` etc. instead of `tasks`, `commands`.

beefham committed:
On launch: data from data.txt is not being read into Duke

Duke is able to write to file but it must also be able to read the file
for the next use

A quick change was made to the constructor of Duke so that the given
file path is read correctly when the GUI launches.

Previously, before the GUI was added, the task list was only set up to read the
data file information when it started run and not when it was created.

Therefore, this fix also makes the run() command in Duke simpler if it
is swapped back to the CLI version, although it should be removed soon.

beefham committed:
change build.gradle to build jar with shadow

beefham committed:
add assertion for number of tasks

beefham committed:
remove unnecessary methods

beefham committed:
added private modifiers and refactor AddTaskCommand to reduce repeated code

beefham committed:
Merge pull request #1 from beefham/branch-Assertions

add assertion for number of tasks

beefham committed:
Merge branch 'master' into branch-CodeQuality

beefham committed:
Merge pull request #2 from beefham/branch-CodeQuality

Branch code quality

beefham committed:
add ability to add tags to tasks

beefham committed:
fix bug related to parsing of data file for tags

beefham committed:
make additions to ui of Duke

beefham committed:
add user guide

beefham committed:
Update README.md
